### PR TITLE
[ion-c-sys] Refactors the traits for `IonCReader` and `IonCWriter`.

### LINKED ISSUE #85: [ion-c-sys] Extract Reader/Writer Traits
We should extract the Reader/Writer traits from `IonCReaderHandle` and `IonCWriterHandle`.  This will make generic/dynamic dispatched code against this API possible, to potentially layer things like Ion Hash (assuming we don't just layer it on `ion-rs`).

### PR DESCRIPTION
**Pull request by almann**
* Adds `IonCAnnotationsFieldWriter` which is a functional way to capture
  the application of annotations and field name.
* Makes `IonCAnnotationsFieldWriterContext` generic and implemented in
  terms of the above.
* Changes `from_buf` to `try_from_buf` to be more consistent in `IonCReaderHandle`.

Resolves #85.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ion-c-sys/src/writer.rs [hunk 1]
```diff
@@ -524,18 +290,308 @@ impl IonCValueWriter for IonCWriterHandle<'_> {
...
+    /// The associated type of value writer when writing within annotations/field context
+    type AFValueWriter: IonCValueWriter;
+
+    /// Writes a value within a context of annotations and/or a field name
+    ///
+    /// Note that it is undefined behavior if a value writing method is **not** called
+    /// or if a value writing method is called **more than once**.  For this reason,
+    /// most users should prefer the `field()` and/or `annotations()` method on the
+    /// [`IonCWriter`](./trait.IonCWriter.html) trait as it doesn't have this edge case.
+    fn write_annotations_and_field<'a, A, F, FN>(
```
Comment (zslayton) [inline ion-c-sys/src/writer.rs:310]: Should `write_annotations_and_field` be marked as `unsafe` given that it's on the caller to understand these requirements to avoid UB?

File: ion-c-sys/src/reader.rs [hunk 2]
```diff
@@ -122,13 +80,7 @@ impl<'a> IonCReaderHandle<'a> {
...
     /// # }
     /// ```
-    #[inline]
-    pub fn depth(&self) -> IonCResult<i32> {
-        let mut depth = 0;
-        ionc!(ion_reader_get_depth(self.reader, &mut depth))?;
-
-        Ok(depth)
-    }
+    fn depth(&self) -> IonCResult<i32>;
```
[the remark this answers is not in the document]
Comment (almann) [reply ion-c-sys/src/reader.rs:83]: They do not--Ion C has a philosophical inclination to rarely use unsigned integers to avoid some of the various arithmetic problems that come from mixing signed and unsigned arithmetic.

File: ion-c-sys/src/reader.rs [hunk 3]
```diff
@@ -462,8 +331,183 @@ impl<'a> IonCReaderHandle<'a> {
...
+        Ok(is_in_struct != 0)
+    }
+
+    #[inline]
+    fn get_field_name(&mut self) -> IonCResult<StrSliceRef> {
+        let mut field = ION_STRING::default();
+        ionc!(ion_reader_get_field_name(self.reader, &mut field))?;
+
+        // make a str slice that is tied to our lifetime
+        let field_str = field.as_str(PhantomData::<&'a u8>::default())?;
```
[the remark this answers is not in the document]
Comment (almann) [reply ion-c-sys/src/reader.rs:426]: The object we're creating is [here][1], the intent is to have the `&str` from the underlying reference holder to be only valid while that reference holder is alive.  The `'a` lifetime is to capture the mutable borrow of the underlying reader while the reference holder is alive.

What is not super crisp to me is if I need to have a per activation `'b` lifetime that ensures that the `&str` returned by `as_ref` and `as_str` don't outlive the reference holder.

[1]: https://github.com/amzn/ion-rust/blob/fc2ed079118cba293f1ef2216da4a9d1bcf3087e/ion-c-sys/src/string.rs#L10-L36

File: ion-c-sys/src/writer.rs [hunk 4]
```diff
@@ -524,18 +290,308 @@ impl IonCValueWriter for IonCWriterHandle<'_> {
...
+        annotations: &'c [&'c str],
+    ) -> IonCAnnotationsFieldWriterContext<'b, 'c, Self> {
+        IonCAnnotationsFieldWriterContext::new_annotations(self, annotations)
+    }
+
+    /// Finalizes writing for the writer and returns the amount of bytes written.
+    fn finish(&mut self) -> IonCResult<usize>;
+}
+
+/// Wrapper over `hWRITER` to make it easier to writers in IonC correctly.
```
[the remark this answers is not in the document]
Comment (almann) [reply ion-c-sys/src/writer.rs]: Fixed

File: ion-c-sys/src/writer.rs [hunk 5]
```diff
@@ -524,18 +290,308 @@ impl IonCValueWriter for IonCWriterHandle<'_> {
...
+///
+/// Specifically supports the `Drop` trait to make sure `ion_writer_close` is run.
+/// Access to the underlying `hWRITER` pointer is done by de-referencing the handle.
+///
+/// See also:
+/// * [IonCWriter](./trait.IonCWriter.html)
+/// * [IonCValueWriter](./trait.IonCValueWriter.html)
+pub struct IonCWriterHandle<'a> {
+    writer: hWRITER,
+    /// Placeholder to tie our lifecycle back to the source of the data--which might not
```
[the remark this answers is not in the document]
Comment (almann) [reply ion-c-sys/src/writer.rs:434]: Let me rephrase, this is obvious copy/paste from the reader.

File: ion-c-sys/src/writer.rs (hunk 1 above)
Comment (almann) [reply ion-c-sys/src/writer.rs:310]: It isn't `unsafe` from the memory model perspective,  just undefined if the user uses the APIs in a weird way.  In other words, there is no leak if someone were to call this method with an empty closure, it is just probably going to lead to a failure because we set up the writing of the field/annotations.

I could've made this API more robust (but chose not to), this code won't leak, but may do weird things if used improperly.  I've added #97 in case we care to make this enhancement.

File: ion-c-sys/src/reader.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (almann) [reply ion-c-sys/src/reader.rs:426]: I did a little test to make sure I understood what is going on here:

This is obviously okay:
```rust
let mut source = String::from("Foobaz");
let data = String::from("Foobar");
{
    let val;
    let slice_ref = StrSliceRef::new(&mut source, data.as_str());
    val = slice_ref.as_str();
    println!("{}", val);
}
```

If I refactor this a bit:

```rust
let mut source = String::from("Foobaz");
let data = String::from("Foobar");
let val;
{
    let slice_ref = StrSliceRef::new(&mut source, data.as_str());
    val = slice_ref.as_str();
}
println!("{}", val);
```

This fails:
```
13 |         val = slice_ref.as_str();
   |               ^^^^^^^^^ borrowed value does not live long enough
14 |     }
   |     - `slice_ref` dropped here while still borrowed
15 |     println!("{}", val);
   |                    --- borrow later used here
```

So we're good.  Any `str` reference borrowed from a `StrSliceRef` is tied to its life, and it mutably borrows a reference to some source and a reference tied to that source for some lifetime 'a that cannot outlive the reader or that reference.

File: ion-c-sys/src/reader.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (almann) [reply ion-c-sys/src/reader.rs:83]: Another place this Ion C idiom pops up in is the various things that return or take `SIZE` which is `i32`, we have a bunch of code that just does `try_into()?` to do those conversions and it is annoying.

File: ion-c-sys/src/reader.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply ion-c-sys/src/reader.rs:426]: Summarizing our offline conversation, I was trying to understand how we would model streaming data (`'reader` outlives `'data_source` outlives `'data_source_slice`) using these constructs. The plan is to model that use case separately, which addresses my concerns.

File: ion-c-sys/src/writer.rs (hunk 1 above)
Comment (zslayton) [reply ion-c-sys/src/writer.rs:310]: > It isn't unsafe from the memory model perspective, just undefined if the user uses the APIs in a weird way.

I'm thinking of "undefined behavior" in the C compiler sense. If it's possible to trigger UB via this API, it should probably be marked `unsafe`. However, it sounds like the undefined behavior being mentioned here would just result in a logic bug, in which case I'm happy to track changing the contract as an issue.
